# Assignment Project - Feature Engineering




## Problem Statement:

The data files Bollywood.csv and Bollywood - 2 .csv contains the box office collection and social media promotion information about the movies released in 2013 – 2015 period. 
.
.
.

Following is the information about the columns (names can vary):

•	Serial Id of movie

•	Release date of movie

•	Release time of movie like Long Weekend (3), Festive Season (1), Holiday Season (2), and Normal (4)

•	Genre of movie like Romance, Thriller, Action, comedy etc.

•	Budget of movie in crores

•	Box office collection of movie in crore

•	YouTube views of movies trailers

•	YouTube likes for the movie trailers

•	YouTube dislikes for the movie trailers


In [1]:
# Import necessary libraries for carrying out the file, mathematical, graphical, etc. operations.

import pandas as pd
from pandas import DataFrame as df
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import numpy as np
import warnings
warnings.filterwarnings("ignore")

Importing movies_listing01 Dataset and looking at the data:

In [2]:
movies_listing01 = pd.read_csv("bollywood.csv")

movies_listing01.head()

No      RelDate           MovieName  ReleaseTime         Genre Budget  \
0   1  Apr-18-2014            2 States            3  RomanceGenre   36Cr   
1   2   Jan-4-2013        Table No. 21            4     Thriller    Cr10   
2   3  Jul-18-2014  Amit Sahni Ki List            4   ComedyGenre   10Cr   
3   4   Jan-4-2013    Rajdhani Express            4    DramaGenre    7Cr   
4   5   Jul-4-2014        Bobby Jasoos           44   ComedyGenre   18Cr   

  BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
0                 104       8576361       26622.0           2527.0   
1                  12       1087320        1129.0            137.0   
2                   4        572336         586.0             54.0   
3                 383         42626          86.0             19.0   
4                10.8       3113427        4512.0           1224.0   

   YoutubeLikeDislikes  
0                29149  
1                 1266  
2                  640  
3                  105  
4                 5736

Importing movies_listing02 Dataset and looking at the data:

In [3]:
movies_listing02 = pd.read_csv("bollywood - 2.csv")

movies_listing02.head()

Movie_ID Release_Date             Movie_Name Release_Time     Genre  \
0        51     4-Jul-14  Lekar Hum Deewana Dil            N   Romance   
1        52    22-Mar-13                  Aatma            N  Thriller   
2        53     4-Apr-14         Main Tera Hero           LW    Comedy   
3        54    22-Mar-13               Sona Spa            N  Thriller   
4        55    29-Mar-13             Himmatwala            N    Action   

   BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0        16                   2.50        3998331           2307   
1         8                   8.78         841891            997   
2        40                  55.00        4903941          12657   
3        10                   0.24          79430             61   
4        50                 100.00        2501277              1   

   Youtube_Dislikes  
0               497  
1               306  
2              2139  
3               112  
4                 1

Looking at both the data sets, we can see that the columns names are different though they signify the same thing. Renaming the columns on the both the datasets to match so that they can be analysed further and used together.

In [4]:
movies_listing01.columns

Index(['No', 'RelDate', 'MovieName', 'ReleaseTime', 'Genre', 'Budget',
       'BoxOfficeCollection', 'YoutubeViews', 'YoutubeLikes',
       'YoutubeDislikes', 'YoutubeLikeDislikes'],
      dtype='object')

In [5]:
movies_listing02.columns = ['No', 'RelDate', 'MovieName', 'ReleaseTime', 'Genre', 'Budget',
       'BoxOfficeCollection', 'YoutubeViews', 'YoutubeLikes',
       'YoutubeDislikes']

movies_listing02.head()

No    RelDate              MovieName ReleaseTime     Genre  Budget  \
0  51   4-Jul-14  Lekar Hum Deewana Dil           N   Romance      16   
1  52  22-Mar-13                  Aatma           N  Thriller       8   
2  53   4-Apr-14         Main Tera Hero          LW    Comedy      40   
3  54  22-Mar-13               Sona Spa           N  Thriller      10   
4  55  29-Mar-13             Himmatwala           N    Action      50   

   BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  
0                 2.50       3998331          2307              497  
1                 8.78        841891           997              306  
2                55.00       4903941         12657             2139  
3                 0.24         79430            61              112  
4               100.00       2501277             1                1

Now these datasets have the same columns. Understanding the elements of each of the datasets:

In [6]:
movies_listing01.isnull().sum()

No                     0
RelDate                0
MovieName              0
ReleaseTime            0
Genre                  0
Budget                 0
BoxOfficeCollection    0
YoutubeViews           0
YoutubeLikes           1
YoutubeDislikes        3
YoutubeLikeDislikes    0
dtype: int64

We see that there are null values in YoutubeDislikes and YoutubeLikes

In [7]:
movies_listing02.isnull().sum()

No                     0
RelDate                0
MovieName              0
ReleaseTime            0
Genre                  0
Budget                 0
BoxOfficeCollection    0
YoutubeViews           0
YoutubeLikes           0
YoutubeDislikes        0
dtype: int64

There are no null values on movies_listing02. So no Additional Validations required.

In [8]:
movies_listing01.dtypes

No                       int64
RelDate                 object
MovieName               object
ReleaseTime              int64
Genre                   object
Budget                  object
BoxOfficeCollection     object
YoutubeViews             int64
YoutubeLikes           float64
YoutubeDislikes        float64
YoutubeLikeDislikes      int64
dtype: object

In [9]:
movies_listing02.dtypes

No                       int64
RelDate                 object
MovieName               object
ReleaseTime             object
Genre                   object
Budget                   int64
BoxOfficeCollection    float64
YoutubeViews             int64
YoutubeLikes             int64
YoutubeDislikes          int64
dtype: object

Changing the columns to appropriate data types across both the datasets

Ideally:
    1. RelDate needs to be a Date Field.
    2. MovieName, Genre are String Values.
    3. Budget, BoxOfficeCollection are Floats.
    4. YoutubeViews, YoutubeLikes, YoutubeDislikes and YoutubeLikeDislikes are all integers.
    5. ReleaseTime is kind of an ennumerator

Working with each of the fields to understand if there are any issues:

## Working with Release Date Field:

Attempting the convert the Date Field into standard values.

Running the below code results in an exception.

movies_listing01['FormattedRelDate'] = pd.to_datetime(movies_listing01['RelDate'])

Doing this we encounter an exception as seen above. This indicates that the string formatting is wrong in the provided dataset. This has to be corrected for all the elements in the dataset.

Note: The above code is Markdowned intentionally to enable the code to run smoothly.

Further understanding the data by splitting the dates to year, month and dates.

In [10]:
movies_listing01['MonthOfRel'] = movies_listing01['RelDate'].str[:3]
movies_listing01['DateOfRel'] = movies_listing01['RelDate'].str[4:6]
movies_listing01['YearOfRel'] = movies_listing01['RelDate'].str[-4:]

In [11]:
movies_listing01.groupby(['DateOfRel'])['No'].count()

DateOfRel
09    1
1-    2
10    1
11    3
12    1
13    1
14    2
15    4
18    6
19    2
2-    2
20    1
21    2
22    2
23    1
24    2
25    2
27    1
28    1
30    1
4-    3
6-    2
7-    3
8-    4
9-    1
Name: No, dtype: int64

Inspecting the dates, they are not having any invalid characters. These need not have any further filteration.

In [12]:
movies_listing01.groupby(['YearOfRel'])['No'].count()

YearOfRel
-204     1
2013    19
2014    31
Name: No, dtype: int64

Looking at the above dataset we see the following incorrectly entered data years:
    1. 204
It should have been 2014 as the data is between 2013-2015.
Identifying the specific sample:    

In [13]:
movies_listing01[movies_listing01.YearOfRel.isin(['-204'])]

No     RelDate MovieName  ReleaseTime    Genre Budget BoxOfficeCollection  \
34  35  Feb-21-204   Highway            4   Drama    30Cr               27.25   

    YoutubeViews  YoutubeLikes  YoutubeDislikes  YoutubeLikeDislikes  \
34       3043012        9100.0            971.0                10071   

   MonthOfRel DateOfRel YearOfRel  
34        Feb        21      -204

Correcting the dataset by replacing the value by a correct value.

In [14]:
movies_listing01.RelDate = movies_listing01.RelDate.replace(to_replace=r'204', value='2014', regex=True)

In [15]:
movies_listing01.groupby(['MonthOfRel'])['No'].count()

MonthOfRel
Apr     1
Aug     1
Feb     9
Jan    10
Jul     6
Jum     1
Jun     3
Mar     7
May     5
Nov     2
Oct     2
Sep     2
Sqp     1
juu     1
Name: No, dtype: int64

Looking at the above dataset we see the following incorrectly entered data months:
    1. Jum
    2. Sqp
    3. juu
    
There are two ways of resolving this data issue:
    1. Assuming the typos based on character sequence and the wrongly placed character. eg: Jum is Jun with 'm' entered instead of 'n'. Similarly Sqp is Sep.
    2. However, we cannot get conclude if juu is Jun ot Jul
    3. Alternatively, we can look at the details of the movie to get the exact information to fill this as this dataset is available publicly.

Since Sqp and Jum can be corrected through (1), we can use that. For (2) I am using internet as the reference to fill the data element.

In [16]:
movies_listing01[movies_listing01.MonthOfRel.isin(['juu'])]

No     RelDate   MovieName  ReleaseTime        Genre Budget  \
17  18  juu-6-2014  Filmistaan            2  ComedyGenre   10Cr   

   BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
17                   6        989096        1608.0            173.0   

    YoutubeLikeDislikes MonthOfRel DateOfRel YearOfRel  
17                 1781        juu        6-      2014

This confirms, filmistaan released in the month of Jun. Making the appropriate corrections across the bollywood dataset

In [17]:
movies_listing01[movies_listing01.MonthOfRel.isin(['Jum'])]

No      RelDate   MovieName  ReleaseTime         Genre Budget  \
12  13  Jum-27-2014  Ek Villain            2  RomanceGenre   35Cr   

   BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
12               105.5       4518748       13558.0           1094.0   

    YoutubeLikeDislikes MonthOfRel DateOfRel YearOfRel  
12                14652        Jum        27      2014

In [18]:
movies_listing01[movies_listing01.MonthOfRel.isin(['Sqp'])]

No      RelDate      MovieName  ReleaseTime        Genre Budget  \
6   7  Sqp-19-2014  Daawat-E-Ishq            4  ComedyGenre   30Cr   

  BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
6                24.6       3905050        8315.0           1373.0   

   YoutubeLikeDislikes MonthOfRel DateOfRel YearOfRel  
6                 9688        Sqp        19      2014

In [19]:
movies_listing01.RelDate = movies_listing01.RelDate.replace(to_replace=r'juu', value='Jun', regex=True)
movies_listing01.RelDate = movies_listing01.RelDate.replace(to_replace=r'Jum', value='Jun', regex=True)
movies_listing01.RelDate = movies_listing01.RelDate.replace(to_replace=r'Sqp', value='Sep', regex=True)
#movies_listing01.RelDate = movies_listing01.RelDate.replace("Sqp-19-2014", "Sep-19-2014")
movies_listing01.head()

No      RelDate           MovieName  ReleaseTime         Genre Budget  \
0   1  Apr-18-2014            2 States            3  RomanceGenre   36Cr   
1   2   Jan-4-2013        Table No. 21            4     Thriller    Cr10   
2   3  Jul-18-2014  Amit Sahni Ki List            4   ComedyGenre   10Cr   
3   4   Jan-4-2013    Rajdhani Express            4    DramaGenre    7Cr   
4   5   Jul-4-2014        Bobby Jasoos           44   ComedyGenre   18Cr   

  BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
0                 104       8576361       26622.0           2527.0   
1                  12       1087320        1129.0            137.0   
2                   4        572336         586.0             54.0   
3                 383         42626          86.0             19.0   
4                10.8       3113427        4512.0           1224.0   

   YoutubeLikeDislikes MonthOfRel DateOfRel YearOfRel  
0                29149        Apr        18      2014  
1                 1266        Jan        4-      2013  
2                  640        Jul        18      2014  
3                  105        Jan        4-      2013  
4                 5736        Jul        4-      2014

These replacements are done, creating another column for storing this formatted date value so that original can be referred to incase of any issues that we might detect.

In [20]:
movies_listing01['FormattedRelDate'] = pd.to_datetime(movies_listing01['RelDate'])

movies_listing01.head()

No      RelDate           MovieName  ReleaseTime         Genre Budget  \
0   1  Apr-18-2014            2 States            3  RomanceGenre   36Cr   
1   2   Jan-4-2013        Table No. 21            4     Thriller    Cr10   
2   3  Jul-18-2014  Amit Sahni Ki List            4   ComedyGenre   10Cr   
3   4   Jan-4-2013    Rajdhani Express            4    DramaGenre    7Cr   
4   5   Jul-4-2014        Bobby Jasoos           44   ComedyGenre   18Cr   

  BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
0                 104       8576361       26622.0           2527.0   
1                  12       1087320        1129.0            137.0   
2                   4        572336         586.0             54.0   
3                 383         42626          86.0             19.0   
4                10.8       3113427        4512.0           1224.0   

   YoutubeLikeDislikes MonthOfRel DateOfRel YearOfRel FormattedRelDate  
0                29149        Apr        18      2014       2014-04-18  
1                 1266        Jan        4-      2013       2013-01-04  
2                  640        Jul        18      2014       2014-07-18  
3                  105        Jan        4-      2013       2013-01-04  
4                 5736        Jul        4-      2014       2014-07-04

Dropping Columns created for validations

In [21]:
movies_listing01 = movies_listing01.drop(['MonthOfRel', 'DateOfRel', 'YearOfRel'], axis = 1)
movies_listing01.head(1)

No      RelDate MovieName  ReleaseTime         Genre Budget  \
0   1  Apr-18-2014  2 States            3  RomanceGenre   36Cr   

  BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
0                 104       8576361       26622.0           2527.0   

   YoutubeLikeDislikes FormattedRelDate  
0                29149       2014-04-18

Applying the same checks on the bollywood2 dataset:

Attempting to format the data field directly to change to a standard format:

In [22]:
movies_listing02['FormattedRelDate'] = pd.to_datetime(movies_listing02['RelDate'])
movies_listing02.head(1)

No   RelDate              MovieName ReleaseTime    Genre  Budget  \
0  51  4-Jul-14  Lekar Hum Deewana Dil           N  Romance      16   

   BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
0                  2.5       3998331          2307              497   

  FormattedRelDate  
0       2014-07-04

We see that the formatting is successful and there are no issues here.

## Working with Genre Field:

Looking at the distribution by grouping them by Genres

In [23]:
movies_listing01.groupby(['Genre'])['No'].count()

Genre
 Drama           1
 DramaGenre     11
Action           1
ActionGenre      9
ComedyGenre     12
RomanceGenre     9
Thriller         5
Thriller         3
Name: No, dtype: int64

In [24]:
movies_listing02.groupby(['Genre'])['No'].count()

Genre
 Drama      23
Action      14
Comedy      24
Romance     17
Thriller    21
Name: No, dtype: int64

The following are the issues that we see:
    1. Drama has a space before the text on both the datasets
    2. Thriller and Action have a trailing space at the end of the text in bollywood dataset
    3. ActionGenre, ComedyGenre and RomanceGenre have 'Genre' appended to the end.
    4. DramaGenre has a leading space and 'Genre' appended to the end.
Removing these irregularities.

In [25]:
movies_listing01.Genre = movies_listing01.Genre.str.strip()
movies_listing01.Genre = movies_listing01.Genre.replace("ActionGenre", "Action")
movies_listing01.Genre = movies_listing01.Genre.replace("ComedyGenre", "Comedy")
movies_listing01.Genre = movies_listing01.Genre.replace("RomanceGenre", "Romance")
movies_listing01.Genre = movies_listing01.Genre.replace("DramaGenre", "Drama")
movies_listing02.Genre = movies_listing02.Genre.str.strip()

Confirming that this has removed the irregularities:

In [26]:
movies_listing01.groupby(['Genre'])['No'].count()

Genre
Action      10
Comedy      12
Drama       12
Romance      9
Thriller     8
Name: No, dtype: int64

In [27]:
movies_listing02.groupby(['Genre'])['No'].count()

Genre
Action      14
Comedy      24
Drama       23
Romance     17
Thriller    21
Name: No, dtype: int64

## Looking at Budget and BoxOfficeCollection:

In [28]:
movies_listing01.groupby(['Budget'])['No'].count()

Budget
100Cr    1
10Cr     4
11Cr     1
120Cr    1
125Cr    1
12Cr     5
15Cr     2
16Cr     1
18Cr     1
20Cr     1
21Cr     1
24Cr     1
27Cr     1
28Cr     1
2Cr      1
30Cr     3
31Cr     1
33Cr     1
35Cr     2
36Cr     2
42Cr     1
48Cr     1
4Cr      2
50Cr     2
5Cr      2
60Cr     1
75Cr     1
7Cr      3
8Cr      1
Cr10     2
Cr30     1
Cr5      1
Cr60     1
Name: No, dtype: int64

In [29]:
movies_listing02.groupby(['Budget'])['No'].count()

Budget
2      2
3      1
4      2
5      6
6      2
8      1
9      1
10     6
11     2
12     4
13     1
14     1
15     8
16     3
18     3
19     1
20     3
21     1
22     2
23     1
24     4
25     4
26     1
27     2
29     1
30     5
31     1
32     1
35     6
40     5
45     1
50     5
60     1
70     1
75     2
85     2
90     2
100    1
115    1
140    1
150    1
Name: No, dtype: int64

Looking Budget we see the following issues:
1. Budget column in bollywood2 has no issues but is all in interger values. This has to be converted to floats.
2. Modifying the bollywood data set to match this result set.

In [30]:
movies_listing02['Budget'] = pd.to_numeric(movies_listing02['Budget'], downcast='float')
movies_listing01['Budget'] = movies_listing01['Budget'].replace(to_replace=r'Cr', value='', regex=True)
movies_listing01['Budget'] = pd.to_numeric(movies_listing01['Budget'], downcast='float')

In [31]:
movies_listing01.groupby(['Budget'])['No'].count()

Budget
2.0      1
4.0      2
5.0      3
7.0      3
8.0      1
10.0     6
11.0     1
12.0     5
15.0     2
16.0     1
18.0     1
20.0     1
21.0     1
24.0     1
27.0     1
28.0     1
30.0     4
31.0     1
33.0     1
35.0     2
36.0     2
42.0     1
48.0     1
50.0     2
60.0     2
75.0     1
100.0    1
120.0    1
125.0    1
Name: No, dtype: int64

In [32]:
movies_listing01.groupby(['BoxOfficeCollection'])['No'].count()

BoxOfficeCollection
1.5       1
1.8       1
10        1
10.25     1
10.8      1
103       1
104       1
105.5     1
11        1
111       1
12        1
12.6      1
129       1
130       1
14.3      1
16        1
162       1
163       1
2         1
2.5       1
21.5      1
22        1
24.6      1
25        1
27        1
27.16     1
27.25     1
3.65      1
32        1
32.7      1
35        1
36        1
377       1
383       2
39.4      1
4         1
40        1
44        1
5.82      1
58.9      1
6         2
62.2      1
63.7      1
65        1
69        1
76.7      1
Cr0.01    1
Cr0.09    1
Cr0.5     1
Name: No, dtype: int64

In [33]:
movies_listing02.groupby(['BoxOfficeCollection'])['No'].count()

BoxOfficeCollection
0.24      1
0.70      1
1.10      2
1.50      1
1.75      1
         ..
300.00    1
311.00    1
340.00    1
395.00    1
735.00    1
Name: No, Length: 84, dtype: int64

There are basically two issues here with BoxOfficeCollection:
    1. Data Type is not consistent. BoxOfficeCollection on bollywood2 should be converted to float.
    2. bollywood dataset has 'cr' appended in some cases which is causing the data to not be in a standard format.

In [34]:
movies_listing02['BoxOfficeCollection'] = pd.to_numeric(movies_listing02['BoxOfficeCollection'], downcast='float')
movies_listing01['BoxOfficeCollection'] = movies_listing01['BoxOfficeCollection'].replace(to_replace=r'Cr', value='', regex=True)
movies_listing01['BoxOfficeCollection'] = pd.to_numeric(movies_listing01['BoxOfficeCollection'], downcast='float')

## Working with YoutubeLikes and YoutubeDislikes:

YoutubeLikes:
1. We know there is a null value in YoutubeLikes only in bollywood dataset. The other dataset is fine.
2. Getting the row where the null value exists:

In [35]:
movies_listing01[movies_listing01.YoutubeLikes.isin([np.nan])]

No      RelDate   MovieName  ReleaseTime   Genre  Budget  \
39  40  Jun-20-2014  Humshakals            2  Comedy    75.0   

    BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
39            63.700001       6280868           NaN              1.0   

    YoutubeLikeDislikes FormattedRelDate  
39                    2       2014-06-20

LikesDislikes contains the sum of Likes and Dislikes for a row. So the value of NaN in Youtube Likes can be obtained from YoutubeLikeDislikes - YoutubeDislikes

In [36]:
movies_listing01['YoutubeLikes'] = movies_listing01['YoutubeLikes'].fillna(movies_listing01['YoutubeLikeDislikes']- movies_listing01['YoutubeDislikes'])
#confirming the null value in YoutubeLikes is replaced
movies_listing01.isnull().sum()

No                     0
RelDate                0
MovieName              0
ReleaseTime            0
Genre                  0
Budget                 0
BoxOfficeCollection    0
YoutubeViews           0
YoutubeLikes           0
YoutubeDislikes        3
YoutubeLikeDislikes    0
FormattedRelDate       0
dtype: int64

Youtube Dislikes:
    1. We know there are 3 null values in Youtube Dislikes in Bollywood Dataset.

Identifying the rows where there are null values:

In [37]:
movies_listing01[movies_listing01.YoutubeDislikes.isin([np.nan])]

No      RelDate           MovieName  ReleaseTime     Genre  Budget  \
9   10  Jan-11-2013           Gangoobai            4     Drama     2.0   
31  32  Feb-15-2013            Murder 3            1  Thriller    12.0   
46  47  Mar-15-2013  Mere Dad Ki Maruti            4    Comedy     5.0   

    BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
9                  0.01          4354           1.0              NaN   
31                25.00       2249055        1879.0              NaN   
46                11.00        569711        1226.0              NaN   

    YoutubeLikeDislikes FormattedRelDate  
9                     2       2013-01-11  
31                 2626       2013-02-15  
46                 1359       2013-03-15

In [38]:
movies_listing01['YoutubeDislikes'] = movies_listing01['YoutubeDislikes'].fillna(movies_listing01['YoutubeLikeDislikes']- movies_listing01['YoutubeLikes'])
#confirming the null value in YoutubeLikes is replaced
movies_listing01.isnull().sum()

No                     0
RelDate                0
MovieName              0
ReleaseTime            0
Genre                  0
Budget                 0
BoxOfficeCollection    0
YoutubeViews           0
YoutubeLikes           0
YoutubeDislikes        0
YoutubeLikeDislikes    0
FormattedRelDate       0
dtype: int64

Identifying any other issues on the Datasets:

In [39]:
movies_listing01.describe()

No  ReleaseTime      Budget  BoxOfficeCollection  YoutubeViews  \
count  51.000000    51.000000   51.000000            51.000000  5.100000e+01   
mean   26.000000     4.137255   27.941177            60.988823  3.504813e+06   
std    14.866069     5.779341   27.754934            91.514565  3.547764e+06   
min     1.000000     1.000000    2.000000             0.010000  4.354000e+03   
25%    13.500000     2.500000   10.000000            10.125000  1.059242e+06   
50%    26.000000     4.000000   18.000000            27.160000  2.375050e+06   
75%    38.500000     4.000000   35.000000            67.000000  5.008115e+06   
max    51.000000    44.000000  125.000000           383.000000  1.898622e+07   

        YoutubeLikes  YoutubeDislikes  YoutubeLikeDislikes  
count      51.000000        51.000000            51.000000  
mean     8607.549020      1265.372549          9872.921569  
std     15343.365202      2084.735916         17190.943468  
min         1.000000         1.000000             2.000000  
25%      1463.500000       150.500000          1683.500000  
50%      3308.000000       556.000000          3773.000000  
75%     11578.500000      1512.000000         12991.500000  
max    101275.000000     10269.000000        111341.000000

In [40]:
movies_listing02.describe()

No      Budget  BoxOfficeCollection  YoutubeViews  \
count   99.000000   99.000000            99.000000  9.900000e+01   
mean   100.000000   30.080809            56.253941  3.258616e+06   
std     28.722813   28.485828           101.429230  3.479723e+06   
min     51.000000    2.000000             0.240000  5.142000e+03   
25%     75.500000   12.000000             8.390000  1.140044e+06   
50%    100.000000   22.000000            29.000000  2.501277e+06   
75%    124.500000   35.000000            55.000000  4.185784e+06   
max    149.000000  150.000000           735.000000  2.317107e+07   

       YoutubeLikes  YoutubeDislikes  
count     99.000000        99.000000  
mean    7445.202020      1170.989899  
std    11187.769162      1722.899736  
min        1.000000         1.000000  
25%     1099.000000       261.500000  
50%     4111.000000       628.000000  
75%     8589.000000      1378.000000  
max    72070.000000     11888.000000

We see that the YoutubeLikeDislikes is missing in bollywood2 dataset. Adding that column:

In [41]:
movies_listing02['YoutubeLikeDislikes'] = movies_listing02['YoutubeDislikes']+ movies_listing02['YoutubeLikes']

## Working with ReleaseTime:

In [42]:
movies_listing01.groupby(['ReleaseTime'])['No'].count()

ReleaseTime
1      3
2     10
3      4
4     33
44     1
Name: No, dtype: int64

In [43]:
movies_listing02.groupby(['ReleaseTime'])['No'].count()

ReleaseTime
FS    14
HS     8
LW    11
N     66
Name: No, dtype: int64

There are essentially two issues here:
    1. An incorrect value in bollywood data set which has a ReleaseTime as 44
    2. The ReleaseTime values are different in the case of both the datasets

In [44]:
movies_listing01.ReleaseTime = movies_listing01.ReleaseTime.replace(44,4)
#Replacing 44 with 4

In [45]:
movies_listing01.groupby(['ReleaseTime'])['No'].count()

ReleaseTime
1     3
2    10
3     4
4    34
Name: No, dtype: int64

From the Problem Statement we understand that:
    1. Release time of movie like Long Weekend (3), Festive Season (1), Holiday Season (2), and Normal (4)
    2. Comparing this with strings with data elements in bollywood2, we can infer that LW = Long Weekend, FS = Festive Season, HS=Holiday Season and N = Normal
    3. The data is now converted to the given Types of ReleaseTimes as described in (1)

Replacing data elements on bollywood2 to match with that of bollywood

In [46]:
movies_listing01.ReleaseTime = movies_listing01.ReleaseTime.replace(1,"Festive Season")
movies_listing01.ReleaseTime = movies_listing01.ReleaseTime.replace(2,"Holiday Season")
movies_listing01.ReleaseTime = movies_listing01.ReleaseTime.replace(3,"Long Weekend")
movies_listing01.ReleaseTime = movies_listing01.ReleaseTime.replace(4,"Normal")
movies_listing02.ReleaseTime = movies_listing02.ReleaseTime.replace('FS',"Festive Season")
movies_listing02.ReleaseTime = movies_listing02.ReleaseTime.replace('HS',"Holiday Season")
movies_listing02.ReleaseTime = movies_listing02.ReleaseTime.replace('LW',"Long Weekend")
movies_listing02.ReleaseTime = movies_listing02.ReleaseTime.replace('N',"Normal")

In [47]:
movies_listing02.groupby(['ReleaseTime'])['No'].count()
#Confirming that both the datasets have the same value sets for ReleaseTime

ReleaseTime
Festive Season    14
Holiday Season     8
Long Weekend      11
Normal            66
Name: No, dtype: int64

## Looking at Youtube Views:

In [48]:
movies_listing01.groupby(['YoutubeViews'])['No'].count() 

YoutubeViews
4354        1
28391       1
42626       1
235767      1
492344      1
569711      1
572336      1
591838      1
764725      1
839270      1
868743      1
989096      1
1041892     1
1076591     1
1077723     1
1087320     1
1176596     1
1200152     1
1551786     1
1741424     1
1829417     1
1888754     1
2249055     1
2333067     1
2343109     1
2375050     1
2397647     1
2435283     1
3043012     1
3113427     1
3503887     1
3905050     1
3998331     1
4005695     1
4042091     1
4518748     1
4687259     1
4823892     1
5192338     1
5454349     1
6144142     1
6280868     1
6336058     1
6395202     1
6604595     1
6841635     1
7048104     1
7641333     1
8576361     1
13798789    1
18986221    1
Name: No, dtype: int64

In [49]:
movies_listing02.groupby(['YoutubeViews'])['No'].count() 

YoutubeViews
5142        1
23186       1
40450       1
68511       1
79430       1
           ..
8788913     1
13270623    1
13279636    1
15889811    1
23171067    1
Name: No, Length: 99, dtype: int64

Looking at the Youtube Views on both the datasets we can see that there are no data quality issues with the dataset.

## Merging the two datasets:

In [50]:
merged_dataset = [movies_listing01, movies_listing02]
dataset = pd.concat(merged_dataset)
dataset.shape

(150, 12)

Validating the data on the merged data set by looking for nulls:

In [51]:
dataset.isnull().sum()

No                     0
RelDate                0
MovieName              0
ReleaseTime            0
Genre                  0
Budget                 0
BoxOfficeCollection    0
YoutubeViews           0
YoutubeLikes           0
YoutubeDislikes        0
YoutubeLikeDislikes    0
FormattedRelDate       0
dtype: int64

There are no null values. Also we see there are 150 rows.

## Checking for Duplicate Values in the dataset

Movie Names cannot be exact duplicates

In [52]:
dataset.shape

(150, 12)

Grouping the movies by name:

In [53]:
distinct_movies = dataset.groupby(['MovieName'])['Budget'].count().reset_index().rename(columns={"Budget":"Occurrences"})

In [54]:
distinct_movies.shape

(149, 2)

We see that there is a difference of one, this would mean that there is one such value which is a duplicate. Sorting the above by descendening and identifying the duplicate record.

In [55]:
distinct_movies.sort_values(by=['Occurrences'], inplace=True, ascending=False)

In [56]:
distinct_movies.head(1)

MovieName  Occurrences
82  Lekar Hum Deewana Dil            2

We can see that there is a duplicate entry in the movie names. Exploring the other other attributes of this movie.

In [57]:
dataset[dataset.MovieName.isin(['Lekar Hum Deewana Dil'])]

No     RelDate              MovieName ReleaseTime    Genre  Budget  \
50  51  Jul-7-2014  Lekar Hum Deewana Dil      Normal  Romance    16.0   
0   51    4-Jul-14  Lekar Hum Deewana Dil      Normal  Romance    16.0   

    BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
50                  2.5       3998331        2307.0            497.0   
0                   2.5       3998331        2307.0            497.0   

    YoutubeLikeDislikes FormattedRelDate  
50                 2804       2014-07-07  
0                  2804       2014-07-04

We see all of the other attributes are a match, thus, removing one of the occurrence.

In [58]:
dataset = dataset.drop_duplicates(subset=['MovieName'])

## Q1. Summary of Issues:
The following are the issues that have been identified and recitified during the data preparation process:
    1. Issues with Release Date:
        a. There are three dates with incorrect month names. These are replaced with the correct names.
        b. There is an year which is incorrectly entered. This is also corrected by replacing it with the correct value.
        c. A formatted date field column  is also created for different types of analytical uses. Original field if leftout with alteration.
    2. Issues with Genre field:
        a. Drama has a space before the text on both the datasets
        b. Thriller and Action have a trailing space at the end of the text in bollywood dataset
        c. ActionGenre, ComedyGenre and RomanceGenre have 'Genre' appended to the end.
        d. DramaGenre has a leading space and 'Genre' appended to the end.
        Leading and Trailing spaces have been removed and where ever 'Genre' is appended to the end, that is removed.
    3. Issues with Budget Field:
        a. Data Type is not consistent across the datasets.
        b. bollywood dataset has 'cr' appended in some cases which is causing the data to not be in a standard format.
    4. Issues with BoxOfficeColleciton Field:
        a. Data Type is not consistent across the datasets.
        b. bollywood dataset has 'cr' appended in some cases which is causing the data to not be in a standard format.
    5. Issues with ReleaseTime:
        a. A value is incorrectly entered in bollywood dataset. This is manually corrected as it is only one value.
        b. The values are not consistent between the two datasets. These have been replaced with the standard values from the Assignment Problem Statement.
    6. Issues with YoutubeLikes Field:
        a. One Null Value in YoutubeLikes in the bollywood dataset. 
        b. This has been rectified by the use of the additional column that is provided as YoutubeLikeDislikes. This column contains the sum of likes and dislikes. Using this information, we have replaced the null value in YoutubeLikes with a value that is equal to YoutubeLikeDislikes - YoutubeDislikes from that specific row.
    7. Issues with YoutubeDislikes Field:
        a. There are three null values in YoutubeDislikes in bollywood dataset. 
        b. These values also can be obtained as a difference between YoutubeLikeDislikes and YoutubeLikes.
    8. Once the data is cleaned on both the datasets and the datasets are merged we see that there is a movie with duplicate entries. One of these is removed.

## Q2: Meta Information of the Dataset

In [59]:
dataset.info()#MetaInfo

<class 'pandas.core.frame.DataFrame'>
Int64Index: 149 entries, 0 to 98
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   No                   149 non-null    int64         
 1   RelDate              149 non-null    object        
 2   MovieName            149 non-null    object        
 3   ReleaseTime          149 non-null    object        
 4   Genre                149 non-null    object        
 5   Budget               149 non-null    float32       
 6   BoxOfficeCollection  149 non-null    float32       
 7   YoutubeViews         149 non-null    int64         
 8   YoutubeLikes         149 non-null    float64       
 9   YoutubeDislikes      149 non-null    float64       
 10  YoutubeLikeDislikes  149 non-null    int64         
 11  FormattedRelDate     149 non-null    datetime64[ns]
dtypes: datetime64[ns](1), float32(2), float64(2), int64(3), object(4)
memory usage: 14.0+ KB


There are a total of 149 entries in the dataset.

## Q3. Movies in each Genre per ReleaseTime

In [60]:
Genre_ReleaseTime = dataset.groupby(['Genre','ReleaseTime'])['Budget'].count().reset_index().rename(columns={"Budget":"Occurrences"})
Genre_ReleaseTime.sort_values(by=['Genre'], inplace=True, ascending=False)
#Genre_ReleaseTime

In [61]:
pivot = Genre_ReleaseTime.pivot_table(index=['Genre'], columns = 'ReleaseTime', values=['Occurrences'], aggfunc='sum')
pivot

Occurrences                                   
ReleaseTime Festive Season Holiday Season Long Weekend Normal
Genre                                                        
Action                   3              3            3     15
Comedy                   3              5            5     23
Drama                    4              6            1     24
Romance                  3              3            4     15
Thriller                 4              1            2     22

The above is the distribution for Movies that are released in each Genre for each of the ReleaseTime.

## Q4. Month of the year where maximum movie releases are seen

In [62]:
dataset['yearOfRelease'], dataset['monthOfRelease'] = dataset['FormattedRelDate'].dt.year,dataset['FormattedRelDate'].dt.month

In [63]:
Month_Year = dataset.groupby(['yearOfRelease','monthOfRelease'])['Budget'].count().reset_index().rename(columns={"Budget":"Occurrences"})
Month_Year_max = Month_Year.loc[Month_Year.groupby(["yearOfRelease"])["Occurrences"].idxmax()]
Month_Year_max.sort_values(by=['Occurrences'], inplace=True, ascending=False)
Month_Year_max

yearOfRelease  monthOfRelease  Occurrences
15           2014               5           12
6            2013               7            9
23           2015               1            5

We can see that the year 2014 and Month 5 (May) has the most number of releases followed by July 2013 with 9 Releases and January 2015 with 5 releases.

## Q5. Month of the year where we see high budgeted movies

High Budgeted movies are the ones with budgets over 25 crs. Creating a column that holds this value.

In [64]:
dataset['HighBudgetMovies'] = np.where((dataset['Budget'] >= 25), 'High Budget', 'Low Budget')

dataset.head(1)

No      RelDate MovieName   ReleaseTime    Genre  Budget  \
0   1  Apr-18-2014  2 States  Long Weekend  Romance    36.0   

   BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
0                104.0       8576361       26622.0           2527.0   

   YoutubeLikeDislikes FormattedRelDate  yearOfRelease  monthOfRelease  \
0                29149       2014-04-18           2014               4   

  HighBudgetMovies  
0      High Budget

Understanding the general spread of High Budget Movies across all years.

In [65]:
Month_Year_Budget = dataset.groupby(['yearOfRelease','monthOfRelease','HighBudgetMovies'])['Budget'].count().reset_index().rename(columns={"Budget":"Occurrences"})
Month_Year_Budget_HighBudget = Month_Year_Budget[Month_Year_Budget.HighBudgetMovies.isin(['High Budget'])]
Month_Year_Budget_HighBudget.sort_values(by=['yearOfRelease','Occurrences'], inplace=True, ascending=False)
Month_Year_Budget_HighBudget['ConcatVal'] = Month_Year_Budget_HighBudget["yearOfRelease"].astype(str)+":"+Month_Year_Budget_HighBudget["Occurrences"].astype(str)

Month_Year_Budget_HighBudget

yearOfRelease  monthOfRelease HighBudgetMovies  Occurrences ConcatVal
43           2015               1      High Budget            3    2015:3
45           2015               2      High Budget            2    2015:2
47           2015               3      High Budget            1    2015:1
22           2014               2      High Budget            4    2014:4
24           2014               3      High Budget            4    2014:4
26           2014               4      High Budget            4    2014:4
20           2014               1      High Budget            3    2014:3
30           2014               6      High Budget            3    2014:3
34           2014               8      High Budget            3    2014:3
40           2014              11      High Budget            3    2014:3
38           2014              10      High Budget            2    2014:2
42           2014              12      High Budget            2    2014:2
28           2014               5      High Budget            1    2014:1
32           2014               7      High Budget            1    2014:1
36           2014               9      High Budget            1    2014:1
11           2013               7      High Budget            5    2013:5
13           2013               8      High Budget            4    2013:4
14           2013               9      High Budget            4    2013:4
2            2013               2      High Budget            3    2013:3
18           2013              11      High Budget            3    2013:3
0            2013               1      High Budget            2    2013:2
4            2013               3      High Budget            2    2013:2
7            2013               5      High Budget            2    2013:2
9            2013               6      High Budget            2    2013:2
16           2013              10      High Budget            2    2013:2

We see that the year 2014 has three months that have the same results. Retaining all the results:

In [66]:
Month_Year_Budget_HighBudget_max = Month_Year_Budget_HighBudget.loc[Month_Year_Budget_HighBudget.groupby(["yearOfRelease"])["Occurrences"].idxmax()]
Month_Year_Budget_HighBudget_max.sort_values(by=['Occurrences'], inplace=True, ascending=False)
Summarized_Budget = Month_Year_Budget_HighBudget_max[['yearOfRelease','Occurrences']]
Summarized_Budget['ConcatVal'] = Summarized_Budget["yearOfRelease"].astype(str)+":"+Summarized_Budget["Occurrences"].astype(str)
Summarized_Budget
budget_result = pd.merge(Summarized_Budget, Month_Year_Budget_HighBudget, on=['ConcatVal'])
budget_result[['yearOfRelease_x','monthOfRelease','Occurrences_y']]

yearOfRelease_x  monthOfRelease  Occurrences_y
0             2013               7              5
1             2014               2              4
2             2014               3              4
3             2014               4              4
4             2015               1              3

We can see that the year 2013 and Month 7 (July) has the most number of high budget releases that is 5 followed by February, Match and April of 2014 with 4 Releases and January 2015 with 3 releases.

## Q6. Top 10 movies with max ROI:

Calculation for ROI:

ROI = (Box Office Collection - Budget)/ Budget

In [67]:
dataset['ROI'] = (dataset['BoxOfficeCollection'] - dataset['Budget'])/dataset['Budget']

dataset.head(1)

No      RelDate MovieName   ReleaseTime    Genre  Budget  \
0   1  Apr-18-2014  2 States  Long Weekend  Romance    36.0   

   BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
0                104.0       8576361       26622.0           2527.0   

   YoutubeLikeDislikes FormattedRelDate  yearOfRelease  monthOfRelease  \
0                29149       2014-04-18           2014               4   

  HighBudgetMovies       ROI  
0      High Budget  1.888889

In [68]:
HighROI  = dataset[['MovieName','FormattedRelDate','ROI']]
HighROI.sort_values(by=['ROI'], inplace=True, ascending=False)
HighROI.head(10)

MovieName FormattedRelDate        ROI
3            Rajdhani Express       2013-01-04  53.714287
14                 Aashiqui 2       2013-04-26   8.166667
39                         PK       2014-12-19   7.647059
82                Grand Masti       2013-09-13   7.514286
85               The Lunchbox       2013-09-20   7.500000
37                     Fukrey       2013-06-14   6.240000
8                    Mary Kom       2014-09-05   5.933333
78                     Shahid       2013-10-18   5.666667
37  Humpty Sharma Ki Dulhania       2014-07-11   5.500000
51         Bhaag Milkha Bhaag       2013-07-12   4.466667